src/Simulation/EntryPointDriver/Maybe.cs

```diff
@@ -1,34 +1,81 @@
-﻿using System;
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
 
 namespace Microsoft.Quantum.EntryPointDriver
 {
+    /// <summary>
+    /// Represents an optional value of type <typeparamref name="T"/>.
+    /// </summary>
+    /// <typeparam name="T">The type of the value.</typeparam>
     internal abstract class Maybe<T>
     {
         private Maybe()
         {
         }
 
+        /// <summary>
+        /// Case analysis on <see cref="Maybe{T}"/>.
+        /// </summary>
+        /// <param name="onNothing">Handles the case when this is <see cref="Nothing"/>.</param>
+        /// <param name="onJust">Handles the case when this is <see cref="Just"/>.</param>
+        /// <typeparam name="TResult">The type of the result.</typeparam>
+        /// <returns>The result of handling the cases.</returns>
         internal abstract TResult Case<TResult>(Func<TResult> onNothing, Func<T, TResult> onJust);
 
+        /// <summary>
+        /// The case of <see cref="Maybe{T}"/> where no value is present.
+        /// </summary>
         internal class Nothing : Maybe<T>
         {
+            /// <summary>
+            /// The singleton instance of <see cref="Nothing"/>.
+            /// </summary>
+            internal static Maybe<T> Instance { get; } = new Nothing();
+
+            private Nothing()
+            {
+            }
+
             internal override TResult Case<TResult>(Func<TResult> onNothing, Func<T, TResult> onJust) => onNothing();
         }
 
+        /// <summary>
+        /// The case of <see cref="Maybe{T}"/> where a value is present.
+        /// </summary>
         internal class Just : Maybe<T>
         {
             private readonly T value;
 
+            /// <summary>
+            /// Creates a <see cref="Just"/> instance.
+            /// </summary>
+            /// <param name="value">The value.</param>
             internal Just(T value) => this.value = value;
 
             internal override TResult Case<TResult>(Func<TResult> onNothing, Func<T, TResult> onJust) => onJust(value);
         }
```

Comment:
Could be useful, but I don't want to add anything that isn't used (especially since this is an internal class). The entry point driver only uses `Case` to extract a value.